File: src/differ_api/services/validators.py
from __future__ import annotations

from pathlib import PurePosixPath

from differ_api.services.errors import ValidationError


def validate_relative_path(path: str) -> str:
    candidate = PurePosixPath(path)
    if not path or candidate.is_absolute() or ".." in candidate.parts:
        raise ValidationError("Invalid path filter")
    return path.rstrip("/")


def validate_path_filters(path_filters: list[str] | None) -> list[str]:
    if not path_filters:
        return []
    normalized: list[str] = []
    for value in path_filters:
        normalized.append(validate_relative_path(value))
    return normalized

File: electron/main/services/validators.ts
import { ValidationError } from '../services/errors'

export function validateRelativePath(path: string): string {
  if (!path) throw new ValidationError('Invalid path')
  if (path.startsWith('/')) throw new ValidationError('Invalid path')
  if (path.includes('..')) throw new ValidationError('Invalid path')
  return path.replace(/\/+$/, '')
}

export function validatePathFilters(filters: string[] | null | undefined): string[] {
  if (!filters || filters.length === 0) return []
  return filters.map((f) => validateRelativePath(f))
}
